dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,374 @@
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_CONVOLUTION_FILTER_GPU_CUH_
+#define DALI_KERNELS_IMGPROC_CONVOLUTION_FILTER_GPU_CUH_
+
+#include <array>
+#include <limits>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "dali/core/cuda_rt_utils.h"
+#include "dali/core/geom/vec.h"
+#include "dali/core/span.h"
+#include "dali/core/static_switch.h"
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/imgproc/convolution/filter_gpu_impl.cuh"
+#include "dali/kernels/imgproc/roi.h"
+#include "dali/kernels/kernel.h"
+
+namespace dali {
+namespace kernels {
+
+/**
+ * @brief Computes the 2D or 3D (``axes_``) convolution of the input (``In``) and the
+ * filter (``W``).
+ *
+ * The input must have the same number of spatial dims as the filter, but
+ * can have extra sequence dim at the beginning (``has_sequence_dim``) and channels
+ * dim at the end (``has_channel_dim``). If the ``enable_roi_`` is false, the outputs and inputs
+ * must have the same size, otherwise they can differ and by passing arbitrary
+ * ``anchors`` to ``Run``, the ROI can be specified.
+ *
+ * For details on the way the convolution is computed, see the docs of the actual cuda kernel.
+ */
+template <typename Out, typename In, typename W, bool has_channel_dim, bool has_sequence_dim,
+          int axes_, bool enable_roi_>
+struct FilterGpu {
+  /* It computes a correlation of the input and the filter.
+  Flip filter in both dimensions for a convolution. */
+
+  static constexpr int axes = axes_;
+  static constexpr bool enable_roi = enable_roi_;
+  static constexpr int ndim =
+      static_cast<int>(has_sequence_dim) + axes + static_cast<int>(has_channel_dim);
+  static constexpr int sequence_dim = has_sequence_dim ? 0 : -1;
+  static constexpr int channels_dim = has_channel_dim ? ndim - 1 : -1;
+  static_assert(axes == 2 || axes == 3);
+  using Intermediate = decltype(std::declval<W>() * std::declval<In>());
+  using StaticConfigT = filter::StaticConfig<axes>;
+  using BlockSetupProviderT = filter::AdaptiveBlock<StaticConfigT>;
+  using BlockSetupT = typename BlockSetupProviderT::BlockSetup;
+  using StaticBlockFactoryT = filter::StaticBlock<StaticConfigT>;
+  using StaticBlockT = typename StaticBlockFactoryT::BlockSetup;
+  using GridSetupT = filter::GridSetup<StaticConfigT>;
+  using InShapeDescT = filter::InShapeDesc<axes>;
+  using OutShapeDescT = filter::OutShapeDesc<axes>;
+  using WorkspaceDescT = filter::WorkspaceDesc<axes>;
+  using SampleDescT = filter::SampleDesc<Out, In, W, Intermediate, axes>;
+
+  void Run(KernelContext& ctx, const TensorListView<StorageGPU, Out, ndim>& out,
+           const TensorListView<StorageGPU, const In, ndim>& in,
+           const TensorListView<StorageGPU, const W, axes>& filters,
+           const span<const ivec<axes>> anchors, boundary::BoundaryType border_type,
```

Comment:
```suggestion
           const span<const ivec<axes>> anchors,
           boundary::BoundaryType border_type,
```
I honestly thought that boundary_type could be per-sample, seeing `border_type` at the end and `span` at the beginning of the line.